x/gamm/types/msgs_test.go

```diff
@@ -867,3 +871,132 @@ func TestMsgExitSwapShareAmountIn(t *testing.T) {
+		},
+		{
+			name: `MsgExitSwapExternAmountOut`,
+			gammMsg: &MsgExitSwapExternAmountOut{
+				Sender:           addr1,
+				PoolId:           1,
+				TokenOut:         coin,
+				ShareInMaxAmount: sdk.NewInt(1),
+			},
+		},
+		{
+			name: "MsgExitPool",
+			gammMsg: &MsgExitPool{
+				Sender:        addr1,
+				PoolId:        1,
+				ShareInAmount: sdk.NewInt(100),
+				TokenOutMins:  sdk.NewCoins(coin),
+			},
+		},
+		{
+			name: "MsgJoinPool",
+			gammMsg: &MsgJoinPool{
+				Sender:         addr1,
+				PoolId:         1,
+				ShareOutAmount: sdk.NewInt(1),
+				TokenInMaxs:    sdk.NewCoins(coin),
+			},
+		},
+		{
+			name: "MsgJoinSwapExternAmountIn",
+			gammMsg: &MsgJoinSwapExternAmountIn{
+				Sender:            addr1,
+				PoolId:            1,
+				TokenIn:           coin,
+				ShareOutMinAmount: sdk.NewInt(1),
+			},
+		},
+		{
+			name: "MsgJoinSwapShareAmountOut",
+			gammMsg: &MsgSwapExactAmountIn{
+				Sender: addr1,
+				Routes: []SwapAmountInRoute{{
+					PoolId:        0,
+					TokenOutDenom: "test",
+				}, {
+					PoolId:        1,
+					TokenOutDenom: "test2",
+				}},
+				TokenIn:           coin,
+				TokenOutMinAmount: sdk.NewInt(1),
+			},
+		},
+		{
+			name: "MsgSwapExactAmountOut",
+			gammMsg: &MsgSwapExactAmountOut{
+				Sender: addr1,
+				Routes: []SwapAmountOutRoute{{
+					PoolId:       0,
+					TokenInDenom: "test",
+				}, {
+					PoolId:       1,
+					TokenInDenom: "test2",
+				}},
+				TokenOut:         coin,
+				TokenInMaxAmount: sdk.NewInt(1),
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var (
+				mockMsgGrant  authz.MsgGrant
+				mockMsgRevoke authz.MsgRevoke
+				mockMsgExec   authz.MsgExec
+			)
+
+			// Authz: Grant Msg
+			typeURL := sdk.MsgTypeURL(tc.gammMsg)
+			grant, err := authz.NewGrant(someDate, authz.NewGenericAuthorization(typeURL), someDate.Add(time.Hour))
+			require.NoError(t, err)
+
+			msgGrant := authz.MsgGrant{Granter: mockGranter, Grantee: mockGrantee, Grant: grant}
+			msgGrantBytes := json.RawMessage(sdk.MustSortJSON(authzcodec.ModuleCdc.MustMarshalJSON(&msgGrant)))
+			err = authzcodec.ModuleCdc.UnmarshalJSON(msgGrantBytes, &mockMsgGrant)
+			require.NoError(t, err)
+
+			// Authz: Revoke Msg
+			msgRevoke := authz.MsgRevoke{Granter: mockGranter, Grantee: mockGrantee, MsgTypeUrl: typeURL}
+			msgRevokeByte := json.RawMessage(sdk.MustSortJSON(authzcodec.ModuleCdc.MustMarshalJSON(&msgRevoke)))
+			err = authzcodec.ModuleCdc.UnmarshalJSON(msgRevokeByte, &mockMsgRevoke)
+			require.NoError(t, err)
+
+			// Authz: Exec Msg
+			msgAny, _ := cdctypes.NewAnyWithValue(tc.gammMsg)
+			msgExec := authz.MsgExec{Grantee: mockGrantee, Msgs: []*cdctypes.Any{msgAny}}
+			execMsgByte := json.RawMessage(sdk.MustSortJSON(authzcodec.ModuleCdc.MustMarshalJSON(&msgExec)))
+			err = authzcodec.ModuleCdc.UnmarshalJSON(execMsgByte, &mockMsgExec)
+			require.NoError(t, err)
+			require.Equal(t, msgExec.Msgs[0].Value, mockMsgExec.Msgs[0].Value)
+		})
```

Comment:
Can we move this all to `apptesting` and call it `TestMessageSerialization` ?

Reply:
Sure !

Reply:
Should I rename package in `msg_test.go` from `types` to `types_test` ? This is to avoid import cycle error.

Reply:
ah yeah, this makes sense to do